src/internal/operators/delayWhen.ts

```diff
@@ -71,151 +80,50 @@ export function delayWhen<T>(delayDurationSelector: (value: T, index: number) =>
-      const destination = this.destination as Subscription;
-      destination.add(notifierSubscription);
-      this.delayNotifierSubscriptions.push(notifierSubscription);
-    }
-  }
-
-  private tryComplete(): void {
-    if (this.completed && this.delayNotifierSubscriptions.length === 0) {
-      this.destination.complete();
-    }
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class SubscriptionDelayObservable<T> extends Observable<T> {
-  constructor(public source: Observable<T>, private subscriptionDelay: Observable<any>) {
-    super();
-  }
-
-  /** @deprecated This is an internal implementation detail, do not use. */
-  _subscribe(subscriber: Subscriber<T>) {
-    this.subscriptionDelay.subscribe(new SubscriptionDelaySubscriber(subscriber, this.source));
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class SubscriptionDelaySubscriber<T> extends Subscriber<T> {
-  private sourceSubscribed: boolean = false;
-
-  constructor(private parent: Subscriber<T>, private source: Observable<T>) {
-    super();
-  }
-
-  protected _next(unused: any) {
-    this.subscribeToSource();
-  }
-
-  protected _error(err: any) {
-    this.unsubscribe();
-    this.parent.error(err);
-  }
-
-  protected _complete() {
-    this.unsubscribe();
-    this.subscribeToSource();
-  }
-
-  private subscribeToSource(): void {
-    if (!this.sourceSubscribed) {
-      this.sourceSubscribed = true;
-      this.unsubscribe();
-      this.source.subscribe(this.parent);
-    }
-  }
+      concat(subscriptionDelay.pipe(take(1), ignoreElements()), source.pipe(delayWhen(delayDurationSelector)));
+  }
+
+  return (source: Observable<T>) =>
+    lift(source, function (this: Subscriber<T>, source: Observable<T>) {
+      const subscriber = this;
+      let index = 0;
+      let isComplete = false;
+      let active = 0;
+      const outerSubscriber = new OperatorSubscriber(
+        subscriber,
+        (value: T) => {
+          const durationNotifier = delayDurationSelector(value, index++);
+          active++;
+          let closed = false;
+          const notify = () => {
+            subscriber.next(value);
+            durationSubscriber?.unsubscribe();
+            !closed && ((closed = true), --active === 0) && isComplete && subscriber.complete();
+          };
+          const durationSubscriber = new OperatorSubscriber(
+            subscriber,
+            notify,
+            undefined,
+            // TODO(benlesh): I'm inclined to say this is _incorrect_ behavior.
+            // A completion should not be a notification. Note the deprecation above
+            notify
+          );
+          durationNotifier.subscribe(durationSubscriber);
+        },
+        undefined,
+        () => {
+          isComplete = true;
+          active === 0 && subscriber.complete();
+          outerSubscriber?.unsubscribe();
```

Comment:
issue: I don't understand why `outerSubscriber?.unsubscribe` needs to be called here. And I cannot see how it could ever require the optional chaining.

Reply:
Yeah, that's unnecessary. Removed.